### ISSUE TITLE
[AIR/Train] Automatically return the framework specific dataset in `train_loop_per_worker`

### ISSUE THREAD

**Issue by amogkam** [ray-project/ray#24974]
### Description

For our DL trainers (Torch, Tensorflow, etc.), users are expected to call `ray.train.get_dataset_shard()` inside their `train_loop_per_worker` function, and then convert that shard to a Torch dataset or TF dataset themselves (via `ds.to_torch()` or `ds.to_tf()`).

Instead, we may want to consider returning the framework specific datasets directly. 

This will allow us to add various optimizations and usability improvements on the Ray Train side automatically for the user rather than having the user implement this themselves.
1. Moving batches to the correct device
2. Pipelining training with moving batches to device
3. Disabling Autosharding for Tensorflow

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
